modules/darwinosl/darwinosl-source-oslog.m

```diff
@@ -0,0 +1,201 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "darwinosl-source-oslog.h"
+#include "messages.h"
+
+/* https://stackoverflow.com/questions/23684727/nsdateformatter-milliseconds-bug */
+static NSString *customBugFixDateFormat = @"yyyy-MM-dd'T'HH:mm:ss'.%06.0f'ZZZZZ";
+
+@implementation NSDateFormatter (BugFix)
+
+- (NSString *) stringFromDateWithMicroseconds:(NSDate *)theDate
+{
+  NSDateComponents *components = [self.calendar components:NSCalendarUnitNanosecond
+                                                fromDate:theDate];
+  double microseconds = lrint((double)components.nanosecond / (double)1000);
+  NSString *dateTimeFormatString = [self stringFromDate:theDate];
+  NSString *formattedString = [NSString stringWithFormat:dateTimeFormatString, microseconds];
+  return formattedString;
+}
+
+@end
+
+@interface OSLogSource ()
+
+@property (strong, nonatomic) OSLogStore *logStore;
+@property (strong, nonatomic) OSLogPosition *logPosition;
+@property (strong, nonatomic) OSLogEnumerator *enumerator;
+@property (assign, nonatomic) OSLogEnumeratorOptions enumeratorOptions;
+@property (strong, nonatomic) NSPredicate *filterPredicate;
+
+@end
+
+@implementation OSLogSource
+
++ (NSDateFormatter *) defaultDateFormatter
+{
+  NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
+  formatter.locale = [NSLocale localeWithLocaleIdentifier:@"en_US_POSIX"];
+  formatter.dateFormat = customBugFixDateFormat;
+  formatter.timeZone = [NSTimeZone localTimeZone];
+  return formatter;
+}
+
++ (NSDateFormatter *) RFC3339DateFormatter
+{
+  static NSDateFormatter *_RFC3339DateFormatter = nil;
+  if (_RFC3339DateFormatter == nil)
+    _RFC3339DateFormatter = OSLogSource.defaultDateFormatter;
+  return _RFC3339DateFormatter;
+}
+
+- (OSLogStore *) openStore
+{
+  g_assert(self.logStore == nil);
+  NSError *error = nil;
+
+  if (@available(macOS 12.0, *))
+    self.logStore = [OSLogStore storeWithScope:OSLogStoreSystem error:&error];
+  else
+    self.logStore = [OSLogStore localStoreAndReturnError:&error];
+
+  if (self.logStore == nil)
+    msg_error("darwinosl: Error getting handle to OSLog store",
+              evt_tag_long("error code", error.code),
+              evt_tag_str("error", error.localizedDescription.UTF8String));
+
+  return self.logStore;
+}
+
+- (OSLogEnumerator *) openEnumeratorWithDate:(NSDate *)startDate
+  filterString:(NSString *)filterString
+  options:(OSLogEnumeratorOptions)options
+{
+  g_assert(self.enumerator == nil);
+  @try
+    {
+      NSPredicate *filterPredicate = (filterString.length > 0 ? [NSPredicate predicateWithFormat:filterString] : nil);
```

Comment:
I might be wrong on this, but looks like that there are some cases when filterString is nil when. Will ObjectiveC survive a `filterString.length` when `filterString==nil`?

Reply:
Same in the catch block as well at  line 117

Reply:
nvm I checked it it is fine

Reply:
yepp, in ObjectiveC the nil is handled safely, automatically (strange, and i did not like it at first, but this is how it works)